interfaces/attrs.go

```diff
@@ -0,0 +1,197 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+type PlugData struct {
+	plug         *snap.PlugInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type SlotData struct {
+	slot         *snap.SlotInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type Attributes interface {
+	SetStaticAttr(key string, value interface{})
+	StaticAttr(key string) (interface{}, error)
+	StaticAttrs() map[string]interface{}
+	Attr(key string) (interface{}, error)
+	Attrs() (map[string]interface{}, error)
+	SetAttr(key string, value interface{}) error
+}
+
+func NewSlotData(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *SlotData {
+	return &SlotData{
+		slot:         slot,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func NewPlugData(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *PlugData {
+	return &PlugData{
+		plug:         plug,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func (attrs *PlugData) Interface() string {
```

Comment:
I am slightly unhappy about the boilerplate duplication between PlugData and SlotData, a lot of the helpers are identical except for attrs.{plug,slot}. I think by re-organizing the structs a bit we could avoid some of this duplication. Here is an outline of what I have in mind: http://paste.ubuntu.com/25416348/ - it avoids duplication but adds a bit of complexity (via the common base type but without losing type safety) so we need think about the trade-off here, i.e. it might not be worth it. With that refactor: Interface,Name,Snap,Apps,StaticAttr,SetStaticAttr,StaticAttrs,Attr,Attrs,SetAttr would all be common code. Only Ref and SecurityTags would be type specific.

Reply:
pushed to https://github.com/mvo5/snappy/tree/hook-attrs-refactor for easier cherry-pick (if we want to go down this path).